### PR TITLE
chore: allow modifying config in data pipelines test setup

### PR DESCRIPTION
**Pull request by mrehan27**
part of: [MBL-215](https://linear.app/customerio/issue/MBL-215/automatic-device-attributes-collection-and-user-agent-updates)

### Changes

- Moved the creation of `AndroidSDKComponent` earlier with `CustomerIOBuilder` initialization so tests don't have to explicitly create it and can reuse it to override dependencies before initializing the module.
- Added `JUnitTest` and `RobolectricTest` classes to avoid rewriting setup code and to make writing both types of tests easier
- Updated `UnitTest` for more flexibility. Tests should avoid directly inheriting it and use `JUnitTest` or `RobolectricTest` instead
- Added `TestConfiguration` to wrap setup parameters in a single class for flexibility
- Added a DSL-like option for convenient initialization of `TestConfiguration`
- Fixed parent classes for tests

### CODE REVIEW

File: datapipelines/src/test/java/io/customer/datapipelines/support/stubs/UnitTestLogger.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        }
+    }
+
+    override fun error(message: String) {
+        runIfMeetsLogLevelCriteria(CioLogLevel.ERROR) {
+            println("[ERROR]: $message")
+        }
+    }
+
+    private fun runIfMeetsLogLevelCriteria(levelForMessage: CioLogLevel, block: () -> Unit) {
```
Comment (Shahroz16) [inline datapipelines/src/test/java/io/customer/datapipelines/support/stubs/UnitTestLogger.kt:32]: too complicated for a test? 😄 maybe let's turn it on by default ? and not log when not needed?

File: datapipelines/src/test/java/io/customer/datapipelines/support/core/TestConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        fun build(): TestConfiguration = TestConfiguration(
+            cdpApiKey = cdpApiKey,
+            sdkConfig = sdkConfig,
+            configurePlugins = configurePlugins
+        )
+
+        companion object {
+            const val TEST_CDP_API_KEY = "TESTING_API_KEY"
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Shahroz16) [reply datapipelines/src/test/java/io/customer/datapipelines/support/core/TestConfiguration.kt:45]: lets just use DSL !! 🪨

**Review by Shahroz16 — APPROVED** [review]
just few comments to make implementation lighter

File: datapipelines/src/test/java/io/customer/datapipelines/support/stubs/UnitTestLogger.kt (hunk 1 above)
Comment (mrehan27) [reply datapipelines/src/test/java/io/customer/datapipelines/support/stubs/UnitTestLogger.kt]: Agreed. But turning on make logs a bit messy. Simplified it more for testing.

**Pull request merged**